### PR TITLE
Fix for Pre-LN Transformers

### PR DESCRIPTION
**Pull request by MaximumEntropy**
This PR fixes bugs in the Pre-LN implementation in transformer encoders and decoders.

### CODE REVIEW

File: nemo/collections/nlp/modules/common/transformer/transformer_encoders.py [hunk 1]
```diff
@@ -61,26 +61,45 @@ def __init__(
...
+        output_states += self_attn_output
+        output_states = self.layer_norm_2(output_states)
 
-        if not self.pre_ln:
-            output_states = self.layer_norm_2(output_states + self_attn_output)
         return output_states
 
+    def forward(self, encoder_query, encoder_mask, encoder_keys):
+        if self.pre_ln:
+            self.forward_preln(encoder_query, encoder_mask, encoder_keys)
```
Comment (mchrzanowski) [inline nemo/collections/nlp/modules/common/transformer/transformer_encoders.py:99]: `return self.forward_preln(...)`

Comment (mchrzanowski) [reply nemo/collections/nlp/modules/common/transformer/transformer_encoders.py:99]: if you can make this an if/else, that would be a bit safer

**Comment by mchrzanowski** [comment]
if the encoders & decoders are basically the same thing, would it make sense to have a single file that they import blocks from?

**Comment by MaximumEntropy** [comment]
> if the encoders & decoders are basically the same thing, would it make sense to have a single file that they import blocks from?

That is already happening to a good extent - both import MultiHeadAttention and PositionWiseFF blocks and that logic lies outside the Encoder/Decoder layers.

File: nemo/collections/nlp/modules/common/transformer/transformer_decoders.py [hunk 2]
```diff
@@ -66,34 +66,52 @@ def __init__(
...
 
         output_states = self.third_sub_layer(enc_dec_attn_output)
+        output_states += enc_dec_attn_output
+        return self.layer_norm_3(output_states)
 
-        if not self.pre_ln:
-            output_states = self.layer_norm_3(output_states + enc_dec_attn_output)
-
-        return output_states
+    # TODO: add Neural Types
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/transformer_decoders.py:109]: Could you please remove this TODO and change inheritance from TransformerDecoderBlock(NeuralModule) to TransformerDecoderBlock(nn.Module).

This is just a block. The actual Neural Module is nemo.collections.nlp.modules.common.decoder_module.DecoderModule

**Review by okuchaiev — COMMENTED** [review]
looks good to me, just one unrelated (to pre_ln) comment.

@RPrenger  does this look OK to you?

File: nemo/collections/nlp/modules/common/transformer/transformer_decoders.py (hunk 2 above)
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/transformer/transformer_decoders.py]: Cool, done

**Comment by RPrenger** [comment]
LGTM

**Review by RPrenger — APPROVED** [review]
LGTM

**Comment by mchrzanowski** [comment]
dont merge yet; i am getting some sort of error:

```
Traceback (most recent call last):
  File "examples/nlp/machine_translation/enc_dec_nmt.py", line 127, in main
    trainer.fit(mt_model)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 513, in fit
    self.dispatch()
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 553, in dispatch
    self.accelerator.start_training(self)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 74, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 111, in start_training
    self._results = trainer.run_train()
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 614, in run_train
    self.run_sanity_check(self.lightning_module)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 863, in run_sanity_check
    _, eval_results = self.run_evaluation(max_batches=self.num_sanity_val_batches)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 732, in run_evaluation
    output = self.evaluation_loop.evaluation_step(batch, batch_idx, dataloader_idx)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/evaluation_loop.py", line 164, in evaluation_step
    output = self.trainer.accelerator.validation_step(args)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/accelerators/accelerator.py", line 178, in validation_step
    return self.training_type_plugin.validation_step(*args)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 290, in validation_step
    return self.model(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/distributed.py", line 671, in forward
    output = self.module(*inputs[0], **kwargs[0])
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/overrides/base.py", line 63, in forward
    output = self.module.validation_step(*inputs, **kwargs)
  File "/home/dcg-adlr-mchrzanowski-source/NeMo2/nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py", line 234, in validation_step
    return self.eval_step(batch, batch_idx, 'val')
  File "/home/dcg-adlr-mchrzanowski-source/NeMo2/nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py", line 199, in eval_step
    log_probs = self(src_ids, src_mask, tgt_ids, tgt_mask)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/dcg-adlr-mchrzanowski-source/NeMo2/nemo/core/classes/common.py", line 721, in __call__
    return wrapped(*args, **kwargs)
  File "/home/dcg-adlr-mchrzanowski-source/NeMo2/nemo/collections/nlp/models/machine_translation/mt_enc_dec_model.py", line 167, in forward
    src_hiddens = self.encoder(src, src_mask)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/dcg-adlr-mchrzanowski-source/NeMo2/nemo/collections/nlp/modules/common/transformer/transformer.py", line 103, in forward
    encoder_hidden_states = self._encoder(embeddings, encoder_mask)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/dcg-adlr-mchrzanowski-source/NeMo2/nemo/collections/nlp/modules/common/transformer/transformer_encoders.py", line 158, in forward
    encoder_states = layer(encoder_states, encoder_attn_mask, memory_states)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/dcg-adlr-mchrzanowski-source/NeMo2/nemo/collections/nlp/modules/common/transformer/transformer_encoders.py", line 99, in forward
    self.forward_preln(encoder_query, encoder_mask, encoder_keys)
  File "/home/dcg-adlr-mchrzanowski-source/NeMo2/nemo/collections/nlp/modules/common/transformer/transformer_encoders.py", line 70, in forward_preln
    encoder_query = self.layer_norm_1(encoder_query)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/normalization.py", line 170, in forward
    input, self.normalized_shape, self.weight, self.bias, self.eps)
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/functional.py", line 2072, in layer_norm
    torch.backends.cudnn.enabled)
TypeError: layer_norm(): argument 'input' (position 1) must be Tensor, not NoneType
```

File: nemo/collections/nlp/modules/common/transformer/transformer_encoders.py (hunk 1 above)
Comment (mchrzanowski) [reply nemo/collections/nlp/modules/common/transformer/transformer_encoders.py:99]: this isnt done in the latest commit. maybe a merge conflict?

**Review by mchrzanowski — CHANGES_REQUESTED** [review]

**Comment by RPrenger** [comment]
Good catch @mchrzanowski .  I would have thought one of our tests would have failed for something like this.

**Comment by MaximumEntropy** [comment]
Nice catch, our CI runs with`pre_ln=false` and this was happening only when `pre_ln=true`. Fixed now.

**Review by mchrzanowski — APPROVED** [review]

**Review by RPrenger — APPROVED** [review]
LGTM

**Pull request merged**